tests/upgrade-tests/utils.py

```diff
@@ -116,3 +121,48 @@ def all_units_ready(model):
     while not all_units_ready(model):
         assert_no_unit_errors(model)
         await asyncio.sleep(1)
+
+async def conjureup(namespace, bundle, channel, snap_channel, model):
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        cmd = 'charm pull --channel=%s cs:~%s/%s %s'
+        cmd %= channel, namespace, bundle, os.path.join(tmpdirname, bundle)
+        cmd = cmd.split()
+        process = await asyncio.create_subprocess_exec(*cmd)
+        await process.wait()
+        assert process.returncode == 0
+        shutil.copytree(
+            os.path.join('/snap/conjure-up/current/spells', bundle),
+            os.path.join(tmpdirname, 'spell')
+        )
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-01_get-kubectl'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-01_get-kubectl.yaml'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-02_cluster-info'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-02_cluster-info.yaml'))
```

Comment:
If this is for automated testing, I'm not sure that prompting for the password would help.  If the test can supply the password if prompted, could it invoke sudo to cache it like conjure-up interactive does?

Alternatively, if the tests are containerized, could they be granted passwordless sudo in a safe way?